File: solutions/cpp/0808-soup-servings.cpp
class Solution {
public:
    double soupServings(int N) {
        if (N > 4451) {
            return 1.0;
        }
        N = (N + 24) / 25;
        std::unordered_map<int, double> memo;

        return dp(N, N, memo);
    }

private:
    double dp(int a, int b, std::unordered_map<int, double>& memo) {
        if (a <= 0 && b <= 0) {
            return 0.5;
        }
        if (a <= 0) {
            return 1.0;
        }
        if (b <= 0) {
            return 0.0;
        }
        int key = a * 5000 + b;
        if (memo.count(key)) {
            return memo[key];
        }
        memo[key] = 0.25 * (dp(a-4, b, memo) + dp(a-3, b-1, memo) + dp(a-2, b-2, memo) + dp(a-1, b-3, memo));
        return memo[key];
    }
};

File: solutions/javascript/0808-soup-servings.js
/**
 * @param {number} n
 * @return {number}
 */
var soupServings = function(N) {
    if (N > 4451) {
        return 1.0;
    }
    N = Math.floor((N + 24) / 25);
    let memo = {};

    return dp(N, N);

    function dp(a, b) {
        if (a <= 0 && b <= 0) {
            return 0.5;
        }
        if (a <= 0) {
            return 1.0;
        }
        if (b <= 0) {
            return 0.0;
        }
        let key = a * 5000 + b;
        if (key in memo) {
            return memo[key];
        }
        memo[key] = 0.25 * (dp(a-4, b) + dp(a-3, b-1) + dp(a-2, b-2) + dp(a-1, b-3));
        return memo[key];
    }
};
